manual/src/refman/extensions/effects.etex

```diff
@@ -0,0 +1,778 @@
+function of type "unit -> unit") and returns a unit to the performer. The
+"Yield" effect is unparameterized and returns a unit when performed. Let us
+consider that a task performing an "Xchg" effect may match with any other task
+also offering to exchange a value.
+
+We shall also define helper functions that simply perform these effects:
+
+\begin{caml_example*}{verbatim}
+let fork f = perform (Fork f)
+let yield () = perform Yield
+let xchg v = perform (Xchg v)
+\end{caml_example*}
+
+A top-level "run" function defines the scheduler:
+
+\begin{caml_example*}{verbatim}
+(* A concurrent round-robin scheduler *)
+let run (main : unit -> unit) : unit =
+  let exchanger = ref None in (* waiting exchanger *)
+  let run_q = Queue.create () in (* scheduler queue *)
+  let enqueue k v =
+    let task () = continue k v in
+    Queue.push task run_q
+  in
+  let dequeue () =
+    if Queue.is_empty run_q then () (* done *)
+    else begin
+      let task = Queue.pop run_q in
+      task ()
+    end
+  in
+  let rec spawn (f : unit -> unit) : unit =
+    match_with f () {
+      retc = dequeue;
+      exnc = (fun e ->
+        print_endline (Printexc.to_string e);
+        dequeue ());
+      effc = fun (type a) (eff : a t) ->
+        match eff with
+        | Yield -> Some (fun (k : (a, unit) continuation) ->
+            enqueue k (); dequeue ())
+        | Fork f -> Some (fun (k : (a, unit) continuation) ->
+            enqueue k (); spawn f)
+        | Xchg n -> Some (fun (k : (int, unit) continuation) ->
+            begin match !exchanger with
+            | Some (n', k') ->
+                exchanger := None; enqueue k' n; continue k n'
+            | None -> exchanger := Some (n, k); dequeue ()
+            end)
+        | _ -> None
+    }
+  in
+  spawn main
+\end{caml_example*}
+
+We use a mutable queue "run_q" to hold the scheduler queue. The FIFO queue
+enables round-robin scheduling of tasks in the scheduler. "enqueue" inserts
+tasks into the queue, and "dequeue" extracts tasks from the queue and runs
+them. The reference cell "exchanger" holds a (suspended) task offering to
+exchange a value. At any time, there is either zero or one suspended task that
+is offering an exchange.
+
+The heavy lifting is done by the "spawn" function. The "spawn" function runs
+the given computation "f" in an effect handler. If "f" returns with a value
+(case "retc"), we dequeue and run the next task from the scheduler queue. If
+the computation "f" raises an exception (case "exnc"), we print the exception
+and run the next task from the scheduler.
+
+The computation "f" may also perform effects. If "f" performs the "Yield"
+effect, the current task is suspended (inserted into the queue of ready tasks),
+and the next task from the scheduler queue is run. If the effect is "Fork f",
+then the current task is suspended, and the new task "f" is executed
+immediately via a tail call to "spawn f". Note that this choice to run the new
```

Comment:
The fact that you are specifying that the call is in the tail position makes the reader wonders about how non tail calls (and stack overflows) are handled inside effect handlers.